qiskit/circuit/equivalence.py

```diff
@@ -62,13 +120,27 @@ def add_equivalence(self, gate, equivalent_circuit):
         _raise_if_param_mismatch(gate.params, equivalent_circuit.parameters)
 
         key = Key(name=gate.name, num_qubits=gate.num_qubits)
-
         equiv = Equivalence(params=gate.params.copy(), circuit=equivalent_circuit.copy())
 
-        if key not in self._map:
-            self._map[key] = Entry(search_base=True, equivalences=[])
+        target = self._set_default_node(key)
+        self._graph[target]["equivs"].append(equiv)
+
+        sources = {
+            Key(name=instruction.operation.name, num_qubits=len(instruction.qubits))
+            for instruction in equivalent_circuit
+        }
+        edges = [
+            (
+                self._set_default_node(source),
+                target,
+                {"index": self._rule_count, "rule": equiv, "source": source},
+            )
+            for source in sources
+        ]
+        self._graph.add_edges_from(edges)
 
-        self._map[key].equivalences.append(equiv)
+        self._num_gates_for_rule[self._rule_count] = len(sources)
+        self._rule_count += 1
```

Comment:
What's the meaning of the integer keys here?  It feels like using `Key` would make more sense rather than closely coupling to the insertion order, but in that case, why don't we just let users infer it from the length of the equivalences output?

Letting it be inferred would also solve the problem that other methods on this class that mutate the library don't sync this property (e.g. `set_entry`).

Reply:
I assume what you are saying is changing this line in `equivalence.py`
```
136                 {"index": self._rule_count, "rule": equiv, "source": source},
```
to something like
```
136                {"key": key, "rule": equiv, "source": source},
```
and changing in basis_translator.py
```
377        self._num_gates_remain_for_rule = dict(num_gates_for_rule)
```
to iterate over `equiv_lib` to create this at runtime for the `BasisTranslator`?

Reply:
See comment above about removing `num_gates_for_rule`.